### DISCUSSION TITLE
Is it possible to get the log written by the logStatements flag?

### DISCUSSION THREAD

**Discussion by madsane29** [simolus3/drift#2034] [Q&A] [answered]
I am trying to implement logging in my app and it would help if I could somehow get the DB logs written by Drift...

![image](https://user-images.githubusercontent.com/40585482/188334173-88b39b8a-93ae-4dc6-9c7d-cf7f0e6a3b41.png)

I write my own logs as well and save it in the DB so the customer can send these logs to me in case something wrong happens...

**Comment by simolus3** [comment, marked as answer]
The easiest way to do this is to override `driftRuntimeOptions.debugPrint` with your custom logging function.

**Reply by madsane29** [reply]
Sounds like a solution but at the moment I have no idea where to override that.

Is there any documentation about it at https://drift.simonbinder.eu/docs/ that I don't find?

Thanks in advance.

**Reply by simolus3** [reply]
It's not documented on the website because it's an API that you rarely need. In addition to https://drift.simonbinder.eu/docs/, I recommend checking the [generated API docs](https://drift.simonbinder.eu/api/) to get a complete overview.

> but at the moment I have no idea where to override that.

You just need to apply it before opening the database, somewhere in `main` will typically do.

**Reply by madsane29** [reply]
By any chance is there some example how am I supposed to override classes like these? I tried in several ways but can't really find the solution.

Thanks in advance.

**Reply by simolus3** [reply]
Reading this again, I think I may have completely misunderstood your question. Do you want to store log messages _in_ a drift database? I'm happy to help you with that, I just assumed you wanted to get logs for SQL statements executed by drift. Sorry for the delay if that's what you meant.

In either way, are you using a logging package at the moment or what does your logging setup look like at the moment? I can show you how to do either thing with drift if you provide more details on how you're processing logs at the moment.

**Reply by madsane29** [reply]
Basically I created a "logs" table in the same database that I use for the app.

In this "logs" table I am writing all kind of logs. Where is the user in the app, did the user create a new object, did the API call happened etc.

So whenever I want to log, I just create a `LogCompanion.insert(...)`, then I make a `logDao.insertEntity()` call.

Now if I could have these `Drift: SELECT * ...` and `Drift: UPDATE table ...` logs that I see in the console whenever I set the `logStatement` true, it would lower the amount of code I have to write to log the inserts/updates/selects what are happening during use of the app. Also could not happen that I miss a place where a DB interaction happens and I forgot to log it.

**Reply by simolus3** [reply]
> So whenever I want to log, I just create a `LogCompanion.insert(...)`, then I make a `logDao.insertEntity()` call.

Nice! Here's a problem though: If you want to log all SQL statements executed by drift and you execute an SQL query for each added log entry, you run into infinite recursion. So if you prefer to log all statements, you'd have to find a way that doesn't create log entries for insertions into your logs table.

Personally, I don't think it makes much sense to log every single database interaction in user apps as long as you log database errors. Still, you can do something like this in `main` or wherever you open your database:

```dart
void main() {
  final myDatabase = AppDatabase();
  driftRuntimeOptions.debugLog = (message) {
    debugPrint(message);
    //  todo: You'll want to be really really sure that this if is working as intended.
    if (!message.contains('INSERT INTO logs')) {
      myDatabase.log.insertOne(LogCompanion.insert(/* use message here */));
    }
  };
}
```

**Reply by madsane29** [reply]
Yeah, it might create a lot of log entities in my DB so might lower the volume somehow later. At the moment it will help me to find some strange bugs in my app. 

This is exactly what I needed, thank you!
